tonic/src/transport/server/mod.rs

```diff
@@ -775,15 +774,11 @@ impl<L> Router<L> {
     /// This method discards any provided [`Server`] TCP configuration.
     ///
     /// [`Server`]: struct.Server.html
-    pub async fn serve_with_incoming<I, IO, IE, ResBody>(
-        self,
-        incoming: I,
-    ) -> Result<(), super::Error>
+    pub async fn serve_with_incoming<I, IO, ResBody>(self, incoming: I) -> Result<(), super::Error>
```

Comment:
We can't add additional constraints in a public interface like this without making a semver-incompatible release, and I don't think we want to do that for this issue.

Reply:
Yeah, I figured that this isn't good.

One "solution" I offered in the description is:

```diff
--- a/tonic/src/transport/server/incoming.rs
+++ b/tonic/src/transport/server/incoming.rs
@@ -27,6 +27,10 @@ where
         let mut incoming = pin!(incoming);
 
         while let Some(item) = incoming.next().await {
+            if item.is_err() {
+                    tracing::debug!(message = "Accept loop error.");
+                    continue
+            }
             yield item.map(ServerIo::new_io)?
         }
     }
```

which will work, but strikes me as too basic.

An alternative would be to change the function to look more like what the TLS enabled version uses (but without the TLS parts). This would make things more complex though:

```rust
#[cfg(not(feature = "tls"))]
pub(crate) fn tcp_incoming<IO, IE>(
    incoming: impl Stream<Item = Result<IO, IE>>,
) -> impl Stream<Item = Result<ServerIo<IO>, crate::Error>>
where
    IO: AsyncRead + AsyncWrite + Unpin + Send + 'static,
    IE: Into<crate::Error>,
{
    async_stream::try_stream! {
        let mut incoming = pin!(incoming);

        let mut tasks = tokio::task::JoinSet::new();

        loop {
            match select(&mut incoming, &mut tasks).await {
                SelectOutput::Incoming(stream) => {
                    yield ServerIo::new_io(stream);
                }

                SelectOutput::Io(io) => {
                    yield io;
                }

                SelectOutput::Err(e) => {
                    tracing::debug!(message = "Accept loop error.", error = %e);
                }

                SelectOutput::Done => {
                    break;
                }
            }
        }
    }
}
```
(untested)

Any other ideas?